### PR TITLE
13554/13776 Edit UI: Add Documents Delivery to miscellaneous filings

### PR DESCRIPTION
**Pull request by lewischenstudio**
*Issue #:* /bcgov/entity#13554 and /bcgov/entity#13776

*Description of changes:*
- editable business contact information
- updated schema for correction
- updated correction filing validation getter
- preload completing party for firm corrections

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/common/YourCompany/BusinessContactInfo.vue [hunk 1]
```diff
@@ -6,7 +6,6 @@
     :hasBusinessContactInfoChange="hasBusinessContactInfoChange"
     :editLabel="editLabel"
     :editedLabel="editSavedLabel"
-    :disableActions="isCorrectionFiling"
```
Comment (severinbeauvais) [inline src/components/common/YourCompany/BusinessContactInfo.vue:9]: Is this for ticket 13776?

Comment (lambert-alex) [reply src/components/common/YourCompany/BusinessContactInfo.vue:9]: Yes it looks like it, I just did it for 13776

File: src/mixins/filing-template-mixin.ts [hunk 2]
```diff
@@ -589,7 +590,7 @@ export default class FilingTemplateMixin extends DateMixin {
     ))
 
     // store current Business Contact
-    this.setBusinessContact({ ...entitySnapshot.authInfo.contact })
+    this.setBusinessContact(filing.correction.contactPoint)
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:593]: The other parsing functions in here get the business contact from the entity snapshot, which I think is OK here, too.
Basically, don't restore the business contact from the contact point object. It's special in that you always want to look at the latest data (not from a draft filing).

File: src/components/common/YourCompany/BusinessContactInfo.vue (hunk 1 above)
Comment (lewischenstudio) [reply src/components/common/YourCompany/BusinessContactInfo.vue:9]: Ticket 13776 overlaps with this ticket, and I mentioned it in planning yesterday.

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:593]: Also this code might fail on the initial correction draft since it probably doesn't contain a contact point object.

Comment (lewischenstudio) [reply src/mixins/filing-template-mixin.ts:593]: There is this problem for contactPoint. We need to send the contactPoint json to `AuthServices.updateContactInfo` API when we update it in the component, which also updates the entitysnapshot. Then it can't tell whether contactPoint is changed or not. So I use `filing.correction` as the reference here.

File: src/views/Correction/FmCorrection.vue [hunk 3]
```diff
@@ -97,6 +103,21 @@ export default class FmCorrection extends Mixins(CommonMixin, FeeMixin, FilingTe
...
+        mailingAddress: {
+          addressCity: mailingAddress?.addressCity,
+          addressCountry: mailingAddress?.addressCountry,
+          addressRegion: mailingAddress?.addressRegion,
+          postalCode: mailingAddress?.postalCode,
+          streetAddress: mailingAddress?.streetAddress,
+          streetAddressAdditional: mailingAddress?.streetAddressAdditional
+        }
+      } as CompletingPartyIF)
+
```
Comment (severinbeauvais) [inline src/views/Correction/FmCorrection.vue:120]: First, I think it's a risky assumption that the completing party's address is the same as the business office.

Second, the other filings that manage completing party data don't do this.

Why are you doing this?

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (lewischenstudio) [reply src/mixins/filing-template-mixin.ts:593]: I can add a condition to use entitysnapshot if it doesn't contain a contact point object, but this contact point is mandatory according to schema. I'm working on it.

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: This is part of the ticket's requirement: "For corrections triggered by clients, need to display the Completing Party, name, email and address, editable, pre-populated with the logged in user values.". Maybe I am missing something here.

File: src/components/common/YourCompany/BusinessContactInfo.vue (hunk 1 above)
Comment (severinbeauvais) [reply src/components/common/YourCompany/BusinessContactInfo.vue:9]: How are you both working on the same ticket? You should always put your name on the tickets you're working on.

Comment (lambert-alex) [reply src/components/common/YourCompany/BusinessContactInfo.vue:9]: I'll close 13776

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: How do the other filings in this UI handle this requirement?

Comment (lambert-alex) [reply src/views/Correction/FmCorrection.vue:120]: Will this not prefill for both client triggered corrections and staff triggered corrections?

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: The other filings in this UI doesn't render pre-populated completing party with the logged in user values.

File: src/components/common/YourCompany/BusinessContactInfo.vue (hunk 1 above)
Comment (lambert-alex) [reply src/components/common/YourCompany/BusinessContactInfo.vue:9]: > How are you both working on the same ticket? You should always put your name on the tickets you're working on, to prevent this.

Different tickets but they had an overlap with each other

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:593]: I think business contact needs to be fetched from snapshot every time -- never from a draft. Otherwise someone could have updated the business contact separately and then this info would be wrong.

PS Watch you don't confuse business contact (object stored in auth db) and contact point (object in the correction filing).

What I noticed here is that this filing handles things differently from the other filings, and that's unexpected.

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: Why doesn't this code prefill things? Maybe it needs to be fixed.

https://github.com/bcgov/business-edit-ui/blob/4b3892439710e413f6e0cb98d29b27d8bd9b8cdb/src/App.vue#L534

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (lewischenstudio) [reply src/mixins/filing-template-mixin.ts:593]: Okay. I will load it from the entity snapshot, but there will be a minor problem when it comes to saving the draft (unless we don't include `this.isCorrectionFiling` in `onContactPointChange()` in `BusinessContactInfo.vue`). This is issue and why I wanted to use the filing.

File: src/components/common/YourCompany/BusinessContactInfo.vue [hunk 4]
```diff
@@ -67,6 +65,7 @@ export default class BusinessContactInfo extends Mixins(CommonMixin) {
       if (
         this.isAlterationFiling ||
         this.isFirmChangeFiling ||
+        this.isCorrectionFiling ||
```
[the remark this answers is not in the document]
Comment (lewischenstudio) [reply src/components/common/YourCompany/BusinessContactInfo.vue:68]: If we include `this.isCorrectionFiling` here, then UI will not detect contact point being changed from resuming a draft.

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:593]: What's the problem with saving the draft?

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: Yeah, this code doesn't load completing party.

Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: Does it work in other filings and is only broken for correction filings? Or only when staff?

File: src/store/getters/state-getters.ts [hunk 5]
```diff
@@ -386,7 +386,8 @@ export const hasCorrectionDataChanged = (state: StateIF): boolean => {
       havePeopleAndRolesChanged(state) ||
       hasShareStructureChanged(state) ||
       areProvisionsRemoved(state) ||
-      haveNewResolutionDates(state)
+      haveNewResolutionDates(state) ||
+      hasContactInfoChanged(state)
```
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:390]: Remove this (and on line 401 below).

Change the business contact info does not create a correction filing.

File: src/mixins/filing-template-mixin.ts (hunk 2 above)
Comment (lewischenstudio) [reply src/mixins/filing-template-mixin.ts:593]: Changing it back. Sorry for the confusion.

File: src/store/getters/state-getters.ts (hunk 5 above)
Comment (lewischenstudio) [reply src/store/getters/state-getters.ts:390]: Removing it.

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: This doesn't work with other filings either. This code doesn't load anything into `setCompletingParty`, and I think we can ask about it during tomorrow's standup.

Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: Does this code work if you're logged in as not staff?

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: Let me try. Hold on.

Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: I don't know where we'd get the name to pre-populate if we are logged in as staff -- a good question to ask tomorrow.

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: I logged in as a client user and got it tested on a change filing. It actually loads the pre-populated values for the completing party component, which is not editable for client users.

Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue:120]: OK, so we need to figure out what we want when logged in as staff (and whether we can do it).

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue:120]: I think I didn't read this line `if (!this.isRoleStaff)` in the code below:
https://github.com/bcgov/business-edit-ui/blob/4b3892439710e413f6e0cb98d29b27d8bd9b8cdb/src/App.vue#L665-L669

You're right. I'll mention it during the standup.

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue]: > Will this not prefill for both client triggered corrections and staff triggered corrections?

The code is now restored/removed, but that code will prefill for either client or staff triggered corrections. We'll see whether we should prefill correction filing for users logged in as staff.

**Comment by lewischenstudio** [comment]
I suggest that we don't merge this ticket until we figure out the prefill condition for completing party, hopefully tomorrow during standup.

File: src/store/getters/state-getters.ts [hunk 6]
```diff
@@ -395,7 +395,6 @@ export const hasCorrectionDataChanged = (state: StateIF): boolean => {
       hasBusinessNameChanged(state) ||
       hasBusinessStartDateChanged(state) ||
       hasNaicsChanged(state) ||
-      hasBusinessStartDateChanged(state) ||
```
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:398]: Why did you remove this?

Comment (lewischenstudio) [reply src/store/getters/state-getters.ts:398]: It's a duplicate from line 396.

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (severinbeauvais) [reply src/views/Correction/FmCorrection.vue]: Wait - "client or staff triggered corrections" can only be done by staff. (I'm not sure if "SBC Staff" will ever do corrections, but if they do then they have non-idir login so this code may inadvertently use their info. This is another question to ask.)

Does this mean we cannot prefill any of this data for a correction?

File: src/store/getters/state-getters.ts (hunk 6 above)
Comment (severinbeauvais) [reply src/store/getters/state-getters.ts:398]: Good catch :)

File: src/views/Correction/FmCorrection.vue (hunk 3 above)
Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue]: We currently cannot prefill any of this for
1. a correction filing
2. any other type of filings, i.e. change filing, if logged in as a staff user.

Comment (lewischenstudio) [reply src/views/Correction/FmCorrection.vue]: It was confirmed from Sarah that from the business point of view, completing party in any type of filing should be left empty if the logged in user is not the owner/account user. So in our case, completing party is not pre-filled for staff users, and the current/original implementation is correct.

File: src/components/common/YourCompany/BusinessContactInfo.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (severinbeauvais) [reply src/components/common/YourCompany/BusinessContactInfo.vue:69]: > For consistency, could you please update YourCompany.vue::showBusinessContactInformation() with the same conditions?

[the remark this answers is not in the document]
Comment (lewischenstudio) [reply src/components/common/YourCompany/BusinessContactInfo.vue:69]: I missed this spot, and it's now fixed.

[the remark this answers is not in the document]
Comment (severinbeauvais) [reply src/components/common/YourCompany/BusinessContactInfo.vue:69]: Thanks, looks good.

File: tests/unit/state-getters.spec.ts [hunk 7]
```diff
@@ -357,6 +363,10 @@ describe('SP/GP correction getters', () => {
       },
       shareStructure: {
         shareClasses: []
+      },
+      authInfo: {
+        contact: {},
+        folioNumber: ''
```
Comment (severinbeauvais) [inline tests/unit/state-getters.spec.ts:369]: Do we still need these changes?

**Review by severinbeauvais — APPROVED** [review]
LGTM

File: tests/unit/state-getters.spec.ts (hunk 7 above)
Comment (lewischenstudio) [reply tests/unit/state-getters.spec.ts]: It was used earlier, but not anymore. They are now reverted.
